### PR TITLE
Change --source-root option to allow omitting sourceRoot

### LINKED ISSUE #1676: Need relative paths in source map
Hi, is there anyway to generate source map without the aboslute path in the file field and without the sourceRoot field?

```
$ traceur --source-maps file --out project.js main.js 
$ cat project.map | jq .
{
  "sourcesContent": [
    "System.registerModule($__placeholder__0, [], $__placeholder__1);",
    "function() {\n        $__placeholder__0\n      }",
    "var __moduleName = $__placeholder__0;",
    "var $__placeholder__0 = $__placeholder__1",
    "class Foo {\n  static hello() {\n    return 'Oh, hi!';\n  }\n}\n\nexport default Foo;\n",
    "($traceurRuntime.createClass)($__placeholder__0, $__placeholder__1, $__placeholder__2)",
    "var $__default = $__placeholder__0",
    "return $__placeholder__0",
    "get $__placeholder__0() { return $__placeholder__1; }",
    "import Foo from 'lib/foo';\n\nconsole.log(Foo.hello());\n",
    "System.get($__placeholder__0)",
    "System.get($__placeholder__0 +'')"
  ],
  "sourceRoot": "/tmp/project/",
  "file": "/tmp/project/project.js",
  "mappings": "AAAA,KAAK,eAAe,AAAC,WAAoB,GAAC,CCA1C,UAAS,AAAD;;ACAR,AAAI,IAAA,CAAA,YAAW,YAAoB,CAAC;ACApC,AAAI,IAAA,MCAJ,SAAM,IAAE,KAIR,ADJwC,CAAA;AEAxC,EAAC,eAAc,YAAY,CAAC,AAAC,WDCpB,KAAI,CAAX,UAAa,AAAD,CAAG;AACb,WAAO,UAAQ,CAAC;IAClB,ECHmF;ACArF,AAAI,IAAA,CAAA,UAAS,EFME,IAAE,AENgB,CAAA;ACAjC,SCAA,aAAwB;AAAE,uBAAwB;IAAE,EDA7B;ANEjB,CDFwD,CAAC;AAA/D,KAAK,eAAe,AAAC,WAAoB,GAAC,CCA1C,UAAS,AAAD;;ACAR,AAAI,IAAA,CAAA,YAAW,YAAoB,CAAC;IOA7B,IAAE,ECAT,CAAA,MAAK,IAAI,AAAC,WAAkB;ADE5B,QAAM,IAAI,AAAC,CAAC,GAAE,MAAM,AAAC,EAAC,CAAC,CAAC;AFFxB,WAAuB;ANEjB,CDFwD,CAAC;AWA/D,KAAK,IAAI,AAAC,CAAC,WAAmB,GAAC,CAAC,CAAA",
  "names": [],
  "sources": [
    "@traceur/generated/TemplateParser/6",
    "@traceur/generated/TemplateParser/5",
    "@traceur/generated/TemplateParser/4",
    "@traceur/generated/TemplateParser/7",
    "lib/foo.js",
    "@traceur/generated/TemplateParser/8",
    "@traceur/generated/TemplateParser/1",
    "@traceur/generated/TemplateParser/3",
    "@traceur/generated/TemplateParser/2",
    "main.js",
    "@traceur/generated/TemplateParser/9",
    "@traceur/generated/TemplateParser/10"
  ],
  "version": 3
}
```

Are there any cases when full path to project's directory is needed inside sourceRoot?

I've tried some hacking, but no luck so far except breaking tests and sourcesContent field :)

### PR DESCRIPTION
**Pull request by johnjbarton**
Now --source-root <false|string|default>.
The default option is as before.

Fixes #1676

### CODE REVIEW

File: src/Compiler.js [hunk 1]
```diff
@@ -158,6 +158,10 @@ export class Compiler {
     tree = this.transform(tree, moduleName);
     // Attach the sourceURL only if the input and output names differ.
     var sourceURL = sourceName !== outputName ? sourceName : undefined;
+    // The sourceRoot argument takes precidence over the option.
+    if (typeof sourceRoot === 'undefined')
```
Comment (arv) [inline src/Compiler.js]: sourceRoot === undefined

File: src/Compiler.js [hunk 2]
```diff
@@ -272,7 +276,16 @@ export class Compiler {
   write(tree, outputName = undefined, sourceRoot = undefined,
       sourceURL = undefined) {
     outputName = this.normalize(outputName);
-    sourceRoot = this.normalize(sourceRoot) || basePath(outputName);
+
+    if (typeof sourceRoot === 'undefined')
```
Comment (arv) [inline src/Compiler.js]: same

File: src/Options.js [hunk 3]
```diff
@@ -287,6 +293,15 @@ export class Options {
           '[false|inline|file|memory], not ' + value);
     }
   }
+
+  get sourceRoot() {
```
Comment (arv) [inline src/Options.js]: Why not use a data property? Do we really need to special case falsy values?

File: src/Options.js [hunk 4]
```diff
@@ -326,6 +342,8 @@ export class Options {
         typeof object.sourceMaps === 'string') {
       this.sourceMaps = object.sourceMaps;
     }
+    if (typeof object.sourceRoot !== 'undefined')
```
Comment (arv) [inline src/Options.js]: object.sourceRoot !== undefined

File: src/Options.js [hunk 5]
```diff
@@ -487,6 +505,14 @@ export function addOptions(flags, commandOptions) {
     'sourceMaps generated to file or inline with data: URL',
     (to) => { return commandOptions.sourceMaps = to; }
   );
+  flags.option('--source-root <false|string|default>',
+    'absolute path to source at execution time',
+    (to) => {
+      if (to==='false')
```
Comment (arv) [inline src/Options.js]: ws

File: src/Options.js [hunk 6]
```diff
@@ -487,6 +505,14 @@ export function addOptions(flags, commandOptions) {
     'sourceMaps generated to file or inline with data: URL',
     (to) => { return commandOptions.sourceMaps = to; }
   );
+  flags.option('--source-root <false|string|default>',
```
Comment (arv) [inline src/Options.js]: Strange value space. Care to expand on how to use this correctly

File: test/unit/node/generated-code-dependencies.js [hunk 7]
```diff
@@ -168,6 +168,68 @@ suite('context test', function() {
     });
   });
 
+  test('compiled modules with --source-maps and sourceroot=/tmp', function(done) {
+    tempFileName = resolve('./out/sourceroot-test.js');
+    var executable = 'node ' + resolve('src/node/command.js');
+    var inputFileName = resolve('test/unit/node/resources/import-x.js');
+    var commandLine = executable + ' --source-maps --source-root=\"/tmp\"' +
```
Comment (arv) [inline test/unit/node/generated-code-dependencies.js]: no need for \" here since you are in a ' string

File: src/Options.js (hunk 6 above)
Comment (johnjbarton) [inline src/Options.js]: We need to express:
  set the sourceRoot to this value,
  leave the sourceRoot undefined,
  set the sourceRoot to the basepath(output file).
 
What shall we use to signal these choices?

Comment (arv) [inline src/Options.js]: Which one maps to false and which one maps to default?

Comment (johnjbarton) [inline src/Options.js]: In this PR currently:
`<string>` set the sourceRoot to this value,
`false` leave the sourceRoot undefined,
`default` set the sourceRoot to the basepath(output file).

Comment (arv) [inline src/Options.js]: What happens when the source root is undefined?

Comment (arv) [inline src/Options.js]: I don't really have a good sense for this. Another option might be to use true|false|&lt;string&gt;.

- `false` for no source root in the generated source map
- `true` for using basepath(output file).
- `"string"` for using the exact string as the source root.

File: src/Compiler.js (hunk 1 above)
Comment (johnjbarton) [inline src/Compiler.js]: done

File: src/Options.js (hunk 3 above)
Comment (johnjbarton) [inline src/Options.js]: done

File: src/Options.js (hunk 4 above)
Comment (johnjbarton) [inline src/Options.js]: done

File: src/Options.js (hunk 6 above)
Comment (johnjbarton) [inline src/Options.js]: done

File: src/Options.js (hunk 5 above)
Comment (johnjbarton) [inline src/Options.js]: done

File: test/unit/node/generated-code-dependencies.js (hunk 7 above)
Comment (johnjbarton) [inline test/unit/node/generated-code-dependencies.js]: done

**Comment by johnjbarton** [comment]
PTAL

File: src/Compiler.js (hunk 1 above)
Comment (arv) [inline src/Compiler.js]: not done?

File: src/Options.js [hunk 8]
```diff
@@ -487,6 +497,17 @@ export function addOptions(flags, commandOptions) {
     'sourceMaps generated to file or inline with data: URL',
     (to) => { return commandOptions.sourceMaps = to; }
   );
+  flags.option('--source-root <true|false|string>',
+    'absolute path to source at execution time. false to omit, ' +
+        'true for directory of output file.',
+    (to) => {
+      if (to === 'false')
+        to = false;
+      if (to === 'true')
```
Comment (arv) [inline src/Options.js]: else if

**Comment by arv** [comment]
I saw that there was a lot of discussion on the bug. Do you feel like this takes us closer to cleaner API?

**Comment by arv** [comment]
My point was: Even if the discussion on the bug has not yet been resolved this might be a step in the right direction and could land before everything was resolved on the issue.

**Comment by johnjbarton** [comment]
PTAL

File: src/Options.js (hunk 8 above)
Comment (arv) [inline src/Options.js]: please fix

**Comment by arv** [comment]
LGTM, one nit

File: src/Options.js (hunk 8 above)
Comment (johnjbarton) [inline src/Options.js]: done

**Pull request closed without merging**